test/Database/LSMTree/ModelIO/Session.hs

```diff
@@ -0,0 +1,75 @@
+{-# LANGUAGE ConstraintKinds          #-}
+{-# LANGUAGE GADTs                    #-}
+{-# LANGUAGE RecordWildCards          #-}
+{-# LANGUAGE ScopedTypeVariables      #-}
+{-# LANGUAGE StandaloneDeriving       #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TupleSections            #-}
+{-# LANGUAGE TypeApplications         #-}
+
+module Database.LSMTree.ModelIO.Session (
+    -- * Sessions
+     Session (..)
+  , newSession
+  , closeSession
+    -- * Internals
+  , new_handle
+  , close_handle
+) where
+
+import           Control.Concurrent.Class.MonadSTM
+import           Control.Monad (void)
+import           Data.Dynamic (Dynamic)
+import           Data.Kind (Type)
+import           Data.Map (Map)
+import qualified Data.Map.Strict as Map
+import           Database.LSMTree.Common (IOLike, SnapshotName)
+
+type Session :: (Type -> Type) -> Type
+data Session m = Session
+  { session_open :: !(TVar m Bool)
+    -- ^ technically we don't need to track whether session is open.
+    -- When session is closed, all the table handles are closed as well,
+    -- So there shouldn't be situation where we have an open table handle
+    -- pointing to closed session.
+    --
+    -- In a model we track this anyway, to catch bugs.
+
+  , snapshots    :: !(TVar m (Map SnapshotName Dynamic))
+    -- ^ model session keeps snapshots in memory.
+
+  , counter      :: !(TVar m Int)
+    -- ^ counter for new handles.
+  , openHandles  :: !(TVar m (Map Int (STM m ())))
+    -- ^ actions to close each open handle
+  }
+
+new_handle :: MonadSTM m => Session m -> TMVar m b -> STM m Int
+new_handle Session {..} ref = do
+    i <- readTVar counter
+    writeTVar counter (i + 1)
+    modifyTVar' openHandles $ Map.insert i $ void $ tryTakeTMVar ref
+    return i
+
+close_handle :: MonadSTM m => Session m -> Int -> STM m ()
+close_handle Session {..} i = do
+    modifyTVar' openHandles $ Map.delete i
```

Comment:
Currently the closing of each handle's reference (by calling `tryTakeTMVar thRef`) is split into two places, once where `close_handle` is called in `close` and again in `closeSession`. So the representation and management of these refs is rather delicately shared between this module and the other module.

I think we could make that a bit cleaner by moving the creation and closing of the `TMVar` into these functions. We'd have `new_handle` be given the initial value for the `TMVar` but then create the `TMVar` itself (returning both the handle id and ref).
```
new_handle :: MonadSTM m => Session m -> b -> STM m (Int, TMVar m b)
new_handle Session {..} refv = do
    i <- readTVar counter
    writeTVar counter (i + 1)
    ref <- newTMVar refv
    let closeref = void $ tryTakeTMVar ref
    modifyTVar' openHandles $ Map.insert i closeref
    return (i, ref)
```
And `close_handle` would also call the closing STM action from the `openHandle` map entry (before deleting the map entry).

Reply:
> And close_handle would also call the closing STM action from the openHandle map entry (before deleting the map entry).

I disagree on this. Currently each handle closing code doesn't depend on session being in a valid state (= contain the handle in it). Your proposal will require that. `tryTakeTMVar` isn't really that complex of a code to duplicate.

If we go for your proposal, then we should add asserts that the handle is recorded in a session closing code.

This all makes me think that *closing session* shouldn't forcefully close the handles. I don't understand why you would need to do that anyway. Closing handles doesn't create snapshots anyway, so it's not required for persistence. Is it only for memory cleanup? In fact the current docs for `closeSession` say:

```
-- This also closes any open table handles in the session. It would typically
-- be good practice however to close all table handles first rather than
-- relying on this for cleanup.
```

So, I'd propose changing it "don't rely on `closeSession` for handle cleanup".

Reply:
Fair enough.

Yes, closing handles is for resource cleanup, both in memory and on-disk.

As for closing the session closing all handles or not, closing the session ought to be able to release the lock file on the session dir (which would allow another process to start using the files) and once we release the lock we can't still be creating new files in the session dir, so we need to prevent any further changes once the session is closed.

Reply:
Table handle operations can check whether their session is open. The model does that in `withModel`, the code checks that both the handle itself and its session are open. I don't see why real implementation couldn't do the same, instead of *relying* on that session is open.

And even so, the creating of new (run) data files is probably better to create with random file names. The only files names which matter are snapshots' (metadata files), and those + their data files shouldn't be modified after creation.